### PR TITLE
Throttler: set timeouts on gRPC communication and on topo communication

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Fixes https://github.com/vitessio/vitess/issues/14164

Thanks to @vmg , the scenario in https://github.com/vitessio/vitess/issues/14164 has been identified. The specific flow is as follows:

- A primary tablet is elected, notices a replica tablet.
- The replica gets decomissioned.
- A new replica tablet (possibly with same alias) is commissioned.

The particular conditions where the error appears are:

- The new replica tablet is assigned a different IP address than the old one, and
- While the old tablet gets decomissioned, any gRPC calls to that tablet (and specifically `CheckThrottler`) _hang_.

The throttler behaves correctly in identifying the change of replicas, and serves requests correctly. However, because its final `CheckThrottler` gRPC call to the decommissioning replica is being _hanged_, the `grpc-go`'s connection is kept open. The specific network environment in https://github.com/vitessio/vitess/issues/14164 never hangs up the connection. Therefore, `grpc-go` keeps a routine maintainance for the connection (`resetConnection`) with some backoff, indefinitely. It then notices the IP address is unreachable and prints a warning message.

In this PR, we expliciitly timeout calls to `tmclient.CheckThrottler()` so that connections are terminated. For good measure (but unrelated to https://github.com/vitessio/vitess/issues/14164), we also timeout calls to the topo server.

## Important notice

This bug is not specific to the throttler. It just so happens that the throttler probes replica tablets at such frequency that it easily falls into the trap. Potentially **any gRPC call made by a tablet to another tablet**, such as in vreplication and Online DDL, is susceptible to this issue.

This PR only addresses the throttler calls.

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/14164

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Review by vmg — APPROVED** [review]
:bow:

**Review by deepthi — APPROVED** [review]
We use timed contexts everywhere in VReplication. You might want to review online DDL to make sure that is true there as well.

**Comment by shlomi-noach** [comment]
> You might want to review online DDL to make sure that is true there as well.

Will do!

File: go/vt/vttablet/tabletserver/throttle/throttler.go [hunk 1]
```diff
@@ -714,6 +714,10 @@ func (throttler *Throttler) Operate(ctx context.Context) {
 
 func (throttler *Throttler) generateTabletHTTPProbeFunction(ctx context.Context, tmClient tmclient.TabletManagerClient, clusterName string, probe *mysql.Probe) (probeFunc func() *mysql.MySQLThrottleMetric) {
 	return func() *mysql.MySQLThrottleMetric {
+		// Some reasonable timeout, to ensure we release connections even if they're hanging (otherwise grpc-go keeps polling those connections forever)
+		ctx, cancel := context.WithTimeout(ctx, 4*mysqlCollectInterval)
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/throttle/throttler.go:718]: Whoops, didn't notice this comment before merging.

The context being passed is not `context.Backgound()`. It is the context sent to `Operate()`. And this is also the reason why disable&enable the throttler fixes the issue: `Disable()` cancels the context, which runs all the way down to cancelling the connection!

[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/throttle/throttler.go:718]: Hiding the original context is fine -- the only consumer of the context is the gRPC call.

**Comment by shlomi-noach** [comment]
v18 backport: https://github.com/vitessio/vitess/pull/14167#pullrequestreview-1655230753

**Pull request merged**